examples/with-nextjs-and-clerk-auth/src/components/ThemeRegistry/RootThemeProvider.tsx

```diff
@@ -1,25 +1,131 @@
 'use client';
 
 import * as React from 'react';
+import { PropsWithChildren, useCallback, useMemo } from 'react';
 
+import { useAuth } from '@clerk/nextjs';
+import { Theme } from '@emotion/react';
 import CssBaseline from '@mui/material/CssBaseline';
+import { createTheme } from '@mui/material/styles';
 import { ThemeProvider } from '@mui/material/styles';
+import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
+import * as themes from '@team-monite/sdk-themes';
+
+import {
+  PrivateMetadata,
+  SelectedTheme,
+  ThemeMode,
+  ThemeVariant,
+} from '@/lib/clerk-api/types';
 
 import NextAppDirEmotionCacheProvider from './EmotionCache';
-import { theme } from './theme';
 
-/**
- * Provides the theme to the Root Application
- * Does not affect the Monite SDK.
- */
-export function RootThemeProvider({ children }: { children: React.ReactNode }) {
+type Context = {
+  setThemeMode: (mode: ThemeMode) => void;
+  setThemeVariant: (theme: ThemeVariant) => void;
+  selectedTheme: SelectedTheme;
+  theme: Theme;
+};
+
+type RootThemeProviderProps = PropsWithChildren<{
+  initialTheme?: SelectedTheme;
+}>;
+
+const QUERY_KEY = ['theme'];
+const METADATA_ROUTE = '/api/user/theme';
+
+export const RootThemeProviderContext = React.createContext<
+  Context | undefined
+>(undefined);
+
+export function RootThemeProvider(props: RootThemeProviderProps) {
+  const { children, initialTheme = ['monite', 'light'] } = props;
+  const { userId } = useAuth();
+
+  const queryClient = useQueryClient();
+
+  const selectedTheme = useQuery<SelectedTheme>({
+    enabled: userId !== null,
+    queryKey: QUERY_KEY,
+    initialData: () => initialTheme,
+    queryFn: async () => {
+      const response = await fetch(METADATA_ROUTE);
+      const { selectedTheme } = (await response.json()) as PrivateMetadata;
+
+      return selectedTheme;
+    },
+  });
+
+  const mutateSelectedTheme = useMutation<SelectedTheme, Error, SelectedTheme>({
+    mutationFn: async (nextSelectedTheme) => {
+      const response = await fetch(METADATA_ROUTE, {
+        method: 'PUT',
+        body: JSON.stringify({ selectedTheme: nextSelectedTheme }),
+        headers: { contentType: 'application/json' },
+      });
+
+      const { selectedTheme } = (await response.json()) as PrivateMetadata;
+
+      return selectedTheme;
+    },
+    onSuccess: (selectedTheme) => {
+      queryClient.setQueryData(QUERY_KEY, selectedTheme);
```

Comment:
I'm glad 🤘 you didn't make an additional `useState`, but work with values directly from `QueryClient`.

But there's one problem with this implementation: until the query runs out - we won't see any changes.

This is the time when optimistic updates come on the scene and the full power of TanStack Query is revealed!

```tsx
  const mutateSelectedTheme = useMutation<
    SelectedTheme,
    Error,
    SelectedTheme,
    SelectedTheme
  >({
    mutationFn(...) { /**/ },
    onMutate: async (nextSelectedTheme) => {
      // 1. Keep previous value
      const previousSelectedTheme =
        queryClient.getQueryData<SelectedTheme>(QUERY_KEY);
      // 2. Set a new value immediately, once you call `.mutate(...)`
      queryClient.setQueryData<SelectedTheme>(QUERY_KEY, nextSelectedTheme);
      // 3. Return previous value as a `Context`
      return previousSelectedTheme;
    },
    onError: (error, variables, previousSelectedTheme) => {
      // 1. Rollback on error
      if (previousSelectedTheme) {
        queryClient.setQueryData<SelectedTheme>(
          QUERY_KEY,
          previousSelectedTheme
        );
      }
      // 2. Notify the user
      console.error(error); // Toast implementation
    },
    onSuccess(...) { /**/ },
  });
```

_Note_ that this case requires specifying a _fourth_ argument for Generic.
```tsx
useMutation<
    SelectedTheme,
    Error,
    SelectedTheme,
    SelectedTheme // Context type
  >
```